### PR TITLE
CUDA: Reduce fork detection overhead.

### PR DESCRIPTION
**Pull request by sklam**
Move fork detection into post-call error handling

### CODE REVIEW

File: numba/cuda/tests/cudapy/test_multiprocessing.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import multiprocessing as mp
+
+import numpy as np
+
+from numba import cuda
+from numba import unittest_support as unittest
+from numba.cuda.testing import skip_on_cudasim
+
+has_mp_get_context = hasattr(mp, 'get_context')
+is_unix = sys.platform.startswith('linux') or sys.platform.startswith('darwin')
```
Comment (pitrou) [inline numba/cuda/tests/cudapy/test_multiprocessing.py]: `os.name == 'posix'` should be more generic :-)

**Comment by pitrou** [comment]
Out of curiosity, for which operations is `get_context()` performance-critical?

**Comment by sklam** [comment]
> for which operations is get_context() performance-critical?

Almost every operation ends up calling get_context() at least once.  Sometimes, an application is required to quickly issue H->D copy, kernel call, D->H copy in multiple asynchronous streams to achieve concurrent kernel calls and copy-execution overlap.  In that case, it is critical to lower the per API call overhead.

There are other cases that require multiple kernel calls to achieve a single task (such as reduction, sorting, searching) due to inter-threadblock communication.  A lower API call overhead reduces the minimum data size to achieve overall speedup.  I have seen it many times that the GPU is starving due to high API call overhead.

File: numba/cuda/tests/cudapy/test_multiprocessing.py (hunk 1 above)
Comment (sklam) [inline numba/cuda/tests/cudapy/test_multiprocessing.py]: updated

**Comment by pitrou** [comment]
Great, thanks.

**Pull request merged**